### ISSUE TITLE
[HOLD for payment 2023-03-13] [$1000] No spacing/margins above and below a Codeblock

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#15299]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Open a chat
2. Send one message - a normal word/sentence, then space and send a codeblock. Then space and send another normal word/sentence.

## Expected Result:
There should be space between the normal messages and codeblock.

## Actual Result:
There's no spacing around the codeblock.

However, [we want an 8 px margin](https://expensify.slack.com/archives/C049HHMV9SM/p1677536946362469?thread_ts=1676670453.605579&cid=C049HHMV9SM) above and below a codeblock.

## Workaround:
unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ x] Android / native
- [ x] Android / Chrome
- [x ] iOS / native
- [x ] iOS / Safari
- [x ] MacOS / Chrome / Safari
- [ x] MacOS / Desktop

**Version Number:** 1.2.74-0
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** 
<img width="1186" alt="image (1)" src="https://user-images.githubusercontent.com/43996225/220148294-aee32d33-401a-4ab2-afab-9fb552897dc4.png">
<img width="960" alt="Untitled 1png" src="https://user-images.githubusercontent.com/43996225/220148329-98e45218-2afd-4309-8941-f8a23b5b2514.png">

**Expensify/Expensify Issue URL:**
**Issue reported by:** @cead22 
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1676670453605579

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~018d44cb79f9fd71df</li>
        <li>Upwork Job ID: 1630342851487846400</li>
        <li>Last Price Increase: 2023-02-27</li>
    </ul>
</details>

**Comment by MelvinBot** [comment]
Triggered auto assignment to @jliexpensify (`Bug`), see https://stackoverflow.com/c/expensify/questions/14418 for more details.

**Comment by MelvinBot** [comment]
### Bug0 Triage Checklist ([Main S/O](https://stackoverflowteams.com/c/expensify/questions/14418))

- [ ] This "bug" occurs on a supported platform (ensure `Platforms` in OP are ✅)
- [ ] This bug is not a duplicate report (check [E/App issues](https://github.com/Expensify/App/issues) and [#expensify-bugs](https://expensify.slack.com/archives/C049HHMV9SM))
    - If it is, comment with a link to the original report, close the issue and add any novel details to the original issue instead
- [ ] This bug is reproducible using the reproduction steps in the OP. [S/O](https://stackoverflowteams.com/c/expensify/questions/14418#:~:text=The%20bug%20is%20reproducible)
    - If the reproduction steps are clear and you're unable to reproduce the bug, check with the reporter and QA first, then close the issue.
    - If the reproduction steps aren't clear and you determine the correct steps, please update the OP.
- [ ] This issue is filled out as thoroughly and clearly as possible
    - Pay special attention to the title, results, platforms where the bug occurs, and if the bug happens on staging/production.
- [ ] I have reviewed and subscribed to the linked Slack conversation to ensure Slack/Github stay in sync

**Comment by jliexpensify** [comment]
Posted a comment here - https://expensify.slack.com/archives/C049HHMV9SM/p1676957142442749?thread_ts=1676670453.605579&cid=C049HHMV9SM

**Comment by jliexpensify** [comment]
[Slack comment here](https://expensify.slack.com/archives/C049HHMV9SM/p1677014573921919?thread_ts=1676670453.605579&cid=C049HHMV9SM) from Bondy, recommending we wait for https://github.com/Expensify/App/pull/14822

**Comment by jliexpensify** [comment]
Alright, I've identified a previous issue and PR ([Slack comment](https://expensify.slack.com/archives/C049HHMV9SM/p1677076219192429?thread_ts=1676670453.605579&cid=C049HHMV9SM)). Asking about the next steps in-thread.

**Comment by jliexpensify** [comment]
Still digging in - https://expensify.slack.com/archives/C049HHMV9SM/p1677452918315599?thread_ts=1676670453.605579&cid=C049HHMV9SM

**Comment by MelvinBot** [comment]
Job added to Upwork: https://www.upwork.com/jobs/~018d44cb79f9fd71df

**Comment by MelvinBot** [comment]
Current assignee @jliexpensify is eligible for the External assigner, not assigning anyone new.

**Comment by MelvinBot** [comment]
Triggered auto assignment to Contributor-plus team member for initial proposal review - @eVoloshchak (`External`)

**Comment by MelvinBot** [comment]
Triggered auto assignment to @mountiny (`External`), see https://stackoverflow.com/c/expensify/questions/7972 for more details.

**Comment by MelvinBot** [comment]
📣 @daraksha-dk! 📣

Hey, it seems we don’t have your contributor details yet! You'll only have to do this once, and this is how we'll hire you on Upwork.
Please follow these steps:

1. Get the email address used to login to your Expensify account. If you don't already have an Expensify account, create one [here](https://new.expensify.com/). If you have multiple accounts (e.g. one for testing), please use your main account email.
2. Get the link to your Upwork profile. It's necessary because we only pay via Upwork. You can access it by logging in, and then clicking on your name. It'll look like [this](https://www.upwork.com/freelancers/~01ca8ad8ce36b9eceb). If you don't already have an account, sign up for one [here](https://www.upwork.com/nx/signup/?dest=home).
3. Copy the format below and paste it in a comment on this issue. Replace the placeholder text with your actual details.

![Screen Shot 2022-11-16 at 4 42 54 PM](https://user-images.githubusercontent.com/30609178/202326034-1a292e64-48fd-4fef-9f4f-7feaaeedfd9f.png)

Format:
```
Contributor details
Your Expensify account email: <REPLACE EMAIL HERE>
Upwork Profile Link: <REPLACE LINK HERE>
```

**Comment by allroundexperts** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Missing spacing above and below the code block.

### What is the root cause of that problem?
The root cause for this is missing margins for the code block.

### What changes do you think we should make in order to solve the problem?

We have to carefully add the margins in order to preserve the tab behaviour. Adding it [here](https://github.com/Expensify/App/blob/7b906c7e9dbfbd9f58f378cc5d1d6f3f2632b407/src/styles/styles.js#L103-L112) will sort of break the tab selection (screen shot attached).
<img width="679" alt="Screenshot 2023-02-28 at 4 33 17 AM" src="https://user-images.githubusercontent.com/30054992/221713300-0b806ac1-0c2f-4aa5-a603-4f2285e2be44.png">

We should add the margins [here](https://github.com/allroundexperts/Expensify/blob/be9b5955305849f9798c31e315d9cf00479895ca/src/components/HTMLEngineProvider/HTMLRenderers/PreRenderer/BasePreRenderer.js#L33) instead. This will preserve the default tab behaviour that we have right now.

Works well after the fix.

<img width="789" alt="Screenshot 2023-02-28 at 4 35 20 AM" src="https://user-images.githubusercontent.com/30054992/221713546-58231720-752a-4514-b82c-a91e24cdc9c4.png">

### What alternative solutions did you explore? (Optional)
None

**Comment by MelvinBot** [comment]
✅ Contributor details stored successfully. Thank you for contributing to Expensify!

**Comment by allroundexperts** [comment]
> > We have to carefully add the margins in order to preserve the tab behaviour. Adding it [here](https://github.com/Expensify/App/blob/7b906c7e9dbfbd9f58f378cc5d1d6f3f2632b407/src/styles/styles.js#L103-L112) will sort of break the tab selection
> 
> I believe displaying outlines for messages is itself a bug, and it's not intentional to show them only for codeblocks within a message. There is no reason to use an outline here as those items are not actionable/interactable. However, since this is ultimately an accessibility issue, we are currently putting it on hold until we can address it properly.
> 
> ### For reference:
> #### Production (NewDot)
> #### Slack's treatment

Thank you for your deep insight @daraksha-dk!

I think this is needed because on mobile, we're making code blocks [pressable](https://github.com/allroundexperts/Expensify/blob/be9b5955305849f9798c31e315d9cf00479895ca/src/components/HTMLEngineProvider/HTMLRenderers/PreRenderer/BasePreRenderer.js#L42).

**Comment by allroundexperts** [comment]
@daraksha-dk That's a good point that you've raised. In my opinion, with the margin applied inside `TouchableWithoutFeedback`, the clickable area of the code block will extend beyond its border which would be a visual bug.

**Comment by allroundexperts** [comment]
@daraksha-dk Another example on desktop. If you long click the code block, on desktop, the context menu opens up. If you long press on the rest of the message nothing happens. Here's a comparison of both approaches on Desktop:

https://user-images.githubusercontent.com/30054992/221895012-783f7850-1fee-4ab8-890a-a8738d150c20.mov

https://user-images.githubusercontent.com/30054992/221895045-0ad285ff-1bf2-4bf2-a17e-28340e924b3a.mov

**Comment by allroundexperts** [comment]
> @daraksha-dk, @allroundexperts, thank you for the thorough testing @allroundexperts's [proposal](https://github.com/Expensify/App/issues/15299#issuecomment-1447284349) looks good to me, we indeed shouldn't break tab selection. Why are we adding styles to `BasePreRenderer`, shouldn't code be rendered by `CodeRenderer`?

@eVoloshchak Thank you for considering my proposal. `CodeRenderer` is using the `InlineCodeBlock` component which does not expose any style property on the container. It just has the `textStyle` property which takes effect somewhere deep. We can also expose a `containerStyles` property in `BasePreRenderer` and add it in `CodeRenderer` if this works for you.

**Comment by eVoloshchak** [comment]
> We can also expose a containerStyles property in BasePreRenderer and add it in CodeRenderer if this works for you.

Let's do that, I think having code styles in `CodeRenderer` is better for readability
@allroundexperts [proposal](https://github.com/Expensify/App/issues/15299#issuecomment-1447284349) looks good to me 
🎀👀🎀 C+ reviewed!
cc: @mountiny

**Comment by MelvinBot** [comment]
📣 @allroundexperts You have been assigned to this job by @mountiny!
Please apply to this job in Upwork and leave a comment on the Github issue letting us know when we can expect a PR to be ready for review 🧑‍💻
Keep in mind: [Code of Conduct](https://github.com/Expensify/App/blob/main/CODE_OF_CONDUCT.md) | [Contributing](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md) 📖

**Comment by mountiny** [comment]
Thanks @eVoloshchak for the review

thanks also to @daraksha-dk for the proposal but going with @allroundexperts feel free to raise a PR! Thanks!

**Comment by allroundexperts** [comment]
PR created.
https://github.com/Expensify/App/pull/15570

**Comment by jliexpensify** [comment]
Hired @eVoloshchak and @allroundexperts

**Comment by mountiny** [comment]
@jliexpensify merged, this should be available for the bonus if no regression

**Comment by MelvinBot** [comment]
`Reviewing` label has been removed, please complete the "BugZero Checklist".

**Comment by MelvinBot** [comment]
The solution for this issue has been :rocket: deployed to production :rocket: in version 1.2.78-0 and is now subject to a 7-day regression period :calendar:. Here is the list of pull requests that resolve this issue:

- https://github.com/Expensify/App/pull/15570


If no regressions arise, payment will be issued on 2023-03-13. :confetti_ball:

After the hold period is over and BZ checklist items are completed, please complete any of the applicable payments for this issue, and check them off once done.
- [ ] External issue reporter
- [ ] Contributor that fixed the issue
- [ ] Contributor+ that helped on the issue and/or PR

As a reminder, here are the bonuses/penalties that should be applied for any External issue:
- Merged PR within 3 business days of assignment - 50% **bonus**
- Merged PR more than 9 business days after assignment - 50% **penalty**

**Comment by MelvinBot** [comment]
**BugZero Checklist:** The PR fixing this issue has been merged! The following checklist ([instructions](https://stackoverflow.com/c/expensify/questions/14574)) will need to be completed before the issue can be closed:

- [ ] [@eVoloshchak / @mountiny] The PR that introduced the bug has been identified. Link to the PR:
- [ ] [@eVoloshchak / @mountiny] The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment:
- [ ] [@eVoloshchak / @mountiny] A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion:
- [ ] [@allroundexperts] Propose [regression test](https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md) steps to ensure the same bug will not reach production again.
- [ ] [@jliexpensify] Link the GH issue for creating/updating the regression test once above steps have been agreed upon:

**Comment by allroundexperts** [comment]
**Regression Test Proposal**

1. Go to https://staging.new.expensify.com/ and open any chat.
2. Enter any comment such that the first line is plain text, second line is a code block and third line is plain text again.
3. Verify that equal amount of margin appears between the lines.

Do we agree 👍 or 👎

**Comment by eVoloshchak** [comment]
- The PR that introduced the bug has been identified. Link to the PR: 

Judjing by [this](https://github.com/Expensify/App/issues/14261#issuecomment-1403903811) comment this was fixed by https://github.com/Expensify/App/pull/14360. I've tested that PR, but it still isn't fixed. 
I think that this [image](https://github.com/Expensify/App/issues/14261#issuecomment-1403903811) contains wrong testing steps. The issue appears if you have a **single** message that contains **both** plain text and code. I might be wrong, but on the image code blocks and plain text are on separate messages. 
I've also tested a couple of old commits that change `BasePreRenderer`, the issue still persisted, so it seems like that's how it was implemented initially.

- A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion: this is a visual bug that could be caught during review, I don't think there are additional checklist items we should add

**Comment by jliexpensify** [comment]
Sorry, I was at a conference the last 3 days - will look at payments now!

**Comment by jliexpensify** [comment]
Paid @allroundexperts and @eVoloshchak

**Comment by jliexpensify** [comment]
I'll get @mountiny to cast his eyes on https://github.com/Expensify/App/issues/15299#issuecomment-1467083970 and I'll whip up that GH

**Comment by jliexpensify** [comment]
@allroundexperts - could you take a look at Vit's feedback and incorporate a screenshot into your Proposal? I'll get that GH sorted after that.

**Comment by allroundexperts** [comment]
@jliexpensify I'm not sure if I am understanding this correctly. Can you please explain what sort of screenshot you're looking for? TIA!

**Comment by mountiny** [comment]
@allroundexperts sorry, I cannot see any screenshot in the comment 
<img width="911" alt="image" src="https://user-images.githubusercontent.com/36083550/226454625-f60f29e4-d171-4f5f-bfca-a44a0dd4aeb6.png">

Basically any screenshot of how the spacing should look like correctly now would be good.

**Comment by mountiny** [comment]
Ahahah, nono, int he regression steps proposal, thank you!

I think this is good to go now @jliexpensify
